### PR TITLE
[CARBONDATA-3804] Provide end-to-end flink integration guide

### PR DESCRIPTION
**Pull request by niuge01**
### Why is this PR needed?
 Provide an end-to-end guide to help user understand and use flink integration module.
 
 ### What changes were proposed in this PR?
 Add file docs/flink-integration-guide.md
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - No

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1248/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2966/

**Comment by Indhumathi27** [comment]
@niuge01 Please add a link in readme.md file

File: docs/flink-integration-guide.md [hunk 1]
```diff
@@ -0,0 +1,193 @@
+##Usage scenarios
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:1]: OK

[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:1]: OK

File: docs/flink-integration-guide.md [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+  Since the flink data written to carbon is endless, in order to ensure the visibility of data 
+  and the controllable amount of data processed during the execution of each insert form stage command, 
+  the user should execute the insert from stage command in a timely manner.
+
+  The execution interval of the insert form stage command should take the data visibility requirements 
+  of the actual business and the flink data traffic. When the data visibility requirements are high 
+  or the data traffic is large, the execution interval should be appropriately shortened.
+
+##Usage description
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:23]: OK

File: docs/flink-integration-guide.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+###Writing process
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:25]: OK

File: docs/flink-integration-guide.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    // 6. Run flink stream.
+    try {
+      environment.execute
+    } catch {
+      case exception: Exception =>
+        // Handle execute exception here.
+    }
+  ```
+
+###Writer properties
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:95]: OK

File: docs/flink-integration-guide.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+####Local Writer
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:97]: OK

File: docs/flink-integration-guide.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+  | Property                             | Name                                 | Description                                                                                             |
+  |--------------------------------------|--------------------------------------|---------------------------------------------------------------------------------------------------------|
+  | CarbonLocalProperty.DATA_TEMP_PATH   | carbon.writer.local.data.temp.path   | Usually is a local path, data will write to temp path first, and mv to target data path finally.        |
+  | CarbonLocalProperty.COMMIT_THRESHOLD | carbon.writer.local.commit.threshold | While written data count reach the threshold, data writer will flush and move data to target data path. | 
+
+####S3 Writer
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:104]: OK

File: docs/flink-integration-guide.md [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  |-----------------------------------|-----------------------------------|---------------------------------------------------------------------------------------------------------|
+  | CarbonS3Property.ACCESS_KEY       | carbon.writer.s3.access.key       | Access key of s3 file system                                                                            |
+  | CarbonS3Property.SECRET_KEY       | carbon.writer.s3.secret.key       | Secret key of s3 file system                                                                            |
+  | CarbonS3Property.ENDPOINT         | carbon.writer.s3.endpoint         | Endpoint of s3 file system                                                                              |
+  | CarbonS3Property.DATA_TEMP_PATH   | carbon.writer.s3.data.temp.path   | Usually is a local path, data will write to temp path first, and mv to target data path finally.        |
+  | CarbonS3Property.COMMIT_THRESHOLD | carbon.writer.s3.commit.threshold | While written data count reach the threshold, data writer will flush and move data to target data path. |
+
+###Insert from stage
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:114]: OK

File: docs/flink-integration-guide.md [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+  [Grammar Description](./dml-of-carbondata.md#insert-data-into-carbondata-table-from-stage-input-files)
+
+##Usage Example Code
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:118]: OK

File: docs/flink-integration-guide.md [hunk 9]
```diff
@@ -0,0 +1,193 @@
+##Usage scenarios
+  
+  A typical scenario is that the data is cleaned and preprocessed by Flink, and then written to Carbon, 
+  for subsequent analysis and queries. 
+
+  The CarbonData flink integration module is used connect Flink and Carbon in the above scenario.
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:6]: OK

File: docs/flink-integration-guide.md [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+      databaseName,
+      tableName,
+      tablePath,
+      tableProperties,
+      writerProperties,
+      carbonProperties
+    )
+     
+    // Build a flink stream and run it.
+    // 1. New a flink execution environment.
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:72]: OK

File: docs/flink-integration-guide.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+  for subsequent analysis and queries. 
+
+  The CarbonData flink integration module is used connect Flink and Carbon in the above scenario.
+
+  The CarbonData flink integration module provides a set of Flink BulkWriter implementations 
+  (CarbonLocalWriter and CarbonS3Writer). The data is processed by the Flink, and finally written into 
+  the stage directory of the target table by the CarbonXXXWriter. 
+
+  By default, those data in table stage directory, can not be immediately queried, those data can be queried 
+  after the "INSERT INTO $tableName STAGE" command is executed.
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:13]: OK

File: docs/flink-integration-guide.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+    }
+  ```
+
+###Writer properties
+
+####Local Writer
+
+  | Property                             | Name                                 | Description                                                                                             |
+  |--------------------------------------|--------------------------------------|---------------------------------------------------------------------------------------------------------|
+  | CarbonLocalProperty.DATA_TEMP_PATH   | carbon.writer.local.data.temp.path   | Usually is a local path, data will write to temp path first, and mv to target data path finally.        |
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:101]: OK

File: docs/flink-integration-guide.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+####S3 Writer
+
+  | Property                          | Name                              | Description                                                                                             |
+  |-----------------------------------|-----------------------------------|---------------------------------------------------------------------------------------------------------|
+  | CarbonS3Property.ACCESS_KEY       | carbon.writer.s3.access.key       | Access key of s3 file system                                                                            |
+  | CarbonS3Property.SECRET_KEY       | carbon.writer.s3.secret.key       | Secret key of s3 file system                                                                            |
+  | CarbonS3Property.ENDPOINT         | carbon.writer.s3.endpoint         | Endpoint of s3 file system                                                                              |
+  | CarbonS3Property.DATA_TEMP_PATH   | carbon.writer.s3.data.temp.path   | Usually is a local path, data will write to temp path first, and mv to target data path finally.        |
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:111]: OK

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2973/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1255/

**Comment by Indhumathi27** [comment]
@niuge01 Please fix the remaining comments

File: docs/flink-integration-guide.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+  | CarbonS3Property.COMMIT_THRESHOLD | carbon.writer.s3.commit.threshold | While written data count reach the threshold, data writer will flush and move data to target data path. |
+
+###Insert from stage
+
+  [Grammar Description](./dml-of-carbondata.md#insert-data-into-carbondata-table-from-stage-input-files)
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md]: OK

File: docs/flink-integration-guide.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+
+    stream.addSink(streamSink)
+
+    try {
+      environment.execute
+    } catch {
+      case exception: Exception =>
+        // TODO
+        throw new UnsupportedOperationException(exception)
+    }
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:212]: OK

File: docs/flink-integration-guide.md [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+
+  The execution interval of the insert form stage command should take the data visibility requirements 
+  of the actual business and the flink data traffic. When the data visibility requirements are high 
+  or the data traffic is large, the execution interval should be appropriately shortened.
+
+##Usage description
+
+###Writing process
+
+  Typical flink stream: Source -> Process -> Output(Carbon Writer Sink)
```
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md:47]: OK

File: docs/flink-integration-guide.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (niuge01) [reply docs/flink-integration-guide.md]: OK

**Comment by niuge01** [comment]
> @niuge01 Please fix the remaining comments

@Indhumathi27 I have fixed, please check it again.

File: docs/flink-integration-guide.md [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+limitations under the License.
+```
+
+-->
+
+# Carbon Flink Integration Guide
+
+## Usage scenarios
+  
+  A typical scenario is that the data is cleaned and preprocessed by Flink, and then written to Carbon,
```
Comment (kunal642) [inline docs/flink-integration-guide.md:25]: Please move this scenario description after introducing the feature.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3002/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1283/

File: docs/flink-integration-guide.md (hunk 17 above)
Comment (niuge01) [reply docs/flink-integration-guide.md]: OK

File: docs/flink-integration-guide.md [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+      case exception: Exception =>
+        // TODO
+        throw new UnsupportedOperationException(exception)
+    }
+  ```
+
+  Insert into table from stage directory.
+  
+  ```sql
+    INSERT INTO test STAGE
```
Comment (Indhumathi27) [inline docs/flink-integration-guide.md:218]: Can create a table test to define the scenario

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1287/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3006/

File: docs/flink-integration-guide.md (hunk 18 above)
Comment (niuge01) [reply docs/flink-integration-guide.md:224]: OK

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1299/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3017/

**Comment by Indhumathi27** [comment]
LGTM

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**